test/boost/big_decimal_test.cc

```diff
@@ -203,3 +203,116 @@ BOOST_AUTO_TEST_CASE(test_big_decimal_sub) {
+    test_cmp(big_decimal("1"), big_decimal("-1"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1.3345E3"), big_decimal("-1.3344E3"), std::strong_ordering::greater);
+    test_cmp(big_decimal("18E2"), big_decimal("2E2"), std::strong_ordering::greater);
+
+    // Test compare with different signs, should hit early return for different signs.
+    test_cmp(big_decimal("-1.3345E3"), big_decimal("1"), std::strong_ordering::less);
+    test_cmp(big_decimal("1.3345E3"), big_decimal("-1.3344E5"), std::strong_ordering::greater);
+
+    // Test compare against 0, should hit early return for different zero.
+    test_cmp(big_decimal("0"), big_decimal("0"), std::strong_ordering::equal);
+    test_cmp(big_decimal("0E6"), big_decimal("0E1"), std::strong_ordering::equal);
+    test_cmp(big_decimal("1"), big_decimal("0"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1"), big_decimal("0E7"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1.32432345E7"), big_decimal("0"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-1"), big_decimal("0"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1"), big_decimal("0E-4"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1.3345E3"), big_decimal("0"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("1"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("187687E3"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("187687E3"), std::strong_ordering::less);
+
+    // Test compare scale-compare algorithm.
+    test_cmp(big_decimal("1.1221310000000023424234E9"), big_decimal("1.267687687968768686"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1.267687687968768686"), big_decimal("1.1221310000000023424234E9"), std::strong_ordering::less);
+    test_cmp(big_decimal("134252342E9"), big_decimal("100023424234E19"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1.1221310000000023424234E9"), big_decimal("-1.267687687968768686"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1.267687687968768686"), big_decimal("-1.1221310000000023424234E9"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-134252342E9"), big_decimal("-100023424234E19"), std::strong_ordering::greater);
+
+    const auto min_scale = std::numeric_limits<int32_t>::min();
+    const auto max_scale = std::numeric_limits<int32_t>::max();
+    const auto huge_unscaled_number = std::numeric_limits<uint64_t>::max();
+
+    // Test compare numbers with huge differences, this used to cause the comparison to allocate large amount of memory and generate stalls.
+    // Reproduces https://github.com/scylladb/scylladb/issues/21716
+    test_cmp(big_decimal("1789679678670986897698780986986E99999999"), big_decimal("1789679678670986897698780986986E-99999999"), std::strong_ordering::greater);
+    test_cmp(big_decimal(max_scale, 1), big_decimal("1"), std::strong_ordering::less);
```

Comment:
I didn't understand the expected ordering until I saw the comment below "note scale has implicit inverted sign"... :-(

Reply:
Yes, it was very confusing to write this test. I kept using the wrong expected outcome.